Improve template tags to support usage outside the loop and add consistency with output vs. return functions

BraadMartin opened the pull request:
I can confirm the PR template works, I just want to be a bit more free form here. :)

I was working on an Alcatraz child theme this weekend and needed to use one of our template tags outside of the loop, but I found that I couldn't do this because all of our template tag functions were only designed to work in the loop.

I decided to go through the template-tags.php file and add support to each function for usage outside of the loop by passing a $post_id param. During this process I made some somewhat opinionated changes including:
- Consistency of comments on functions
- Consistency of alcatraz_some_thing() and alcatraz_the_some_thing()
- Moved output of .entry-footer inside of the alcatraz_entry_footer() function to match out alcatraz_entry_header function, and updated all our templates
- Reworked our alcatraz_edit_post_link() function to be a bit more custom, and moved it to alcatraz_entry_footer() so that it outputs in the same place consistently
- Started passing $post_id to all of our template tag filters
- Added `alcatraz_entry_meta_post_types` and `alcatraz_entry_footer_post_types` filters to control output of these things at the post type level

I'm hoping that these changes apply cleanly to any existing Alcatraz site's that have been built, but I'd love scrutiny from both of you guys, @jgonzo127 and @carrieforde, and testing on any sites that you have going. I think making our template tag functions more flexible like this is definitely a good idea, and I needed this for my project today, but we should only merge this once you guys have reviewed it and are happy with it.

BraadMartin committed:
Update alcatraz_posted_on() template tag to accept a post_id, and add alcatraz_get_posted_on

BraadMartin committed:
Update alcatraz_entry_meta to support passing a post_id and return instead of echo

BraadMartin committed:
Update alcatraz_entry_header and alcatraz_entry_title to accept a post_id, and make comments more consistent

BraadMartin committed:
Update alcatraz_entry_footer and alcatraz_edit_post_link to accept a post_id, and move all edit link output to entry_footer

BraadMartin committed:
Add a capability check to the edit link template tag and add comment about future todo in alcatraz_entry_footer

BraadMartin committed:
One more grunt build to update the translation files

BraadMartin committed:
Fix bug in entry header output logic

BraadMartin commented:
Just added another commit to this to fix a bug I had introduced in my reworking of the entry_title output logic, which I'm remembering now we already went a few rounds on. It _should_ now be correct. :)

carrieforde commented:
This isn't working me. I tested this on a local copy of [Clayton's site](http://www.claytongerard.com/), and the blog post loop breaks (it loads the first post only, and it only loads the content, not the title and meta), and the titles aren't loading in his single portfolio anymore.

I also tested this on the blog on my personal site's conversion to Alcatraz. Same thing, it breaks the blog loop, but this time, nothing loads except the header for the first post.

I won't merge this PR until the loops stop breaking.

BraadMartin committed:
Switch output of entry header content to a hook for easier customization

BraadMartin commented:
Just updated the entry_header functions to use a new hook `alcatraz_entry_header_inside` that should give us much more flexibility for hooking new output into the entry header. I still need to add a similar treatment for entry_footer but I ran out of time this morning. I'll try to get that in tonight.

BraadMartin committed:
Overhaul entry footer template tag into an action call and introduce new template tags

BraadMartin committed:
Remove unneeded CSS

BraadMartin committed:
Update translation file

BraadMartin commented:
Alright, after a big final push on the entry footer and associated template tags, this PR is definitely ready for your review @carrieforde and @jgonzo127, and if it looks good then I'd say it's ready for merge.

Here's the updated list of everything that has changed in this PR:
- All template tag functions that should accept a post_id now do
- All template tag functions now return if they are `alcatraz_thing()` and echo if `alcatraz_the_thing()`
- The function that outputs the edit post link has been renamed `alcatraz_the_edit_post_link()`, and a corresponding function `alcatraz_edit_post_link()` that returns the link has been added
- Both `alcatraz_entry_header` and `alcatraz_entry_footer()` have had their guts converted to a simple hook call (`alcatraz_entry_header_inside` and `alcatraz_entry_footer_inside`), which allows for turning off the default output or adding your own output anywhere inside the wrapping elements
- The <footer> wrap element in `alcatraz_entry_footer()` was moved to the function from the template to match `alcatraz_entry_header()`
- The comments link in alcatraz_entry_footer that had an unfortunate implementation in core and couldn't be easily converted for our purposes has been removed (I wrote more about this in Slack)
- Introduced `alcatraz_entry_meta_post_types` to control which post types get the entry_meta output
- Renamed `alcatraz_get_the_social_network_icons()` to `alcatraz_get_social_network_icons()`
- Moved `alcatraz_get_social_networks()` to theme-options.php (now it is with our other get_\* functions)
- Renamed `alcatraz_get_the_sub_page_nav()` to `alcatraz_get_sub_page_nav()`
- Added `alcatraz_get_taxonomy_term_list()` and `alcatraz_the_taxonomy_term_list()` functions as our "Output term links in the given taxonomy from the given post as a simple list" functions (these are pretty neat)
- Switched to using new taxonomy_term_list functions in place of `get_the_category_list()` and `get_the_tag_list()`
- Removed `alcatraz_categorized_blog()` and `alcatraz_category_transient_flusher()` silliness
- Updated template tag function calls through our template files to use new functions where appropriate

I've tested this thing and in almost every case it should just work. The specific cases where we might need to update existing child themes would be:
- Any child theme templates that call old template tags that now return instead of echo (`alcatraz_entry_title()`, `alcatraz_entry_footer()`) 
- Any child theme templates that call template tag functions that no longer exist (`alcatraz_edit_post()`, `alcatraz_get_the_sub_page_nav()`, `alcatraz_get_the_social_network_icons()`)

I believe that should do it, but please both of you review this, ask any questions, point out where things could be better, etc. These are the kinds of changes we won't want to make in the future when we have a ton of sites built on Alcatraz. Although making them now is going to cause some compatibility work on existing projects, if we think this puts us on a better path to maintain backwards compatibility into the future, I think it's the right thing to do, and the sooner the better.

carrieforde commented on inc/template-tags.php:
```diff
@@ -28,245 +37,237 @@ function alcatraz_posted_on() {
 	$byline = sprintf(
 		esc_html_x( 'by %s', 'post author', 'alcatraz' ),
-		'<span class="author vcard"><a class="url fn n" href="' . esc_url( get_author_posts_url( get_the_author_meta( 'ID' ) ) ) . '">' . esc_html( get_the_author() ) . '</a></span>'
+		'<span class="author vcard"><a class="url fn n" href="' . esc_url( get_author_posts_url( $author_id ) ) . '">' . esc_html( get_the_author_meta( 'nicename', $author_id ) ) . '</a></span>'
```
Right now, this is showing the username, not the user's full name, which in most cases, I think would be what we'd want.

BraadMartin replied:
Good call, let's try display_name there. Each user can set their display_name on their User Profile screens.

carrieforde commented on inc/template-tags.php:
```diff
@@ -28,245 +37,237 @@ function alcatraz_posted_on() {
-function alcatraz_category_transient_flusher() {
-	if ( defined( 'DOING_AUTOSAVE' ) && DOING_AUTOSAVE ) {
-		return;
+function alcatraz_entry_footer( $post_id = 0 ) {
```
I like it.

carrieforde commented on inc/theme-hooks.php:
```diff
@@ -123,6 +123,53 @@ function alcatraz_output_sub_page_nav() {
+
+	alcatraz_the_edit_post_link( $post_id );
+
+	echo '<hr>';
```
Let's nix this...it borders on too opinionated, and I don't want to rewrite the footer just to remove it.

BraadMartin replied:
:+1:

carrieforde commented on inc/theme-hooks.php:
```diff
@@ -123,6 +123,53 @@ function alcatraz_output_sub_page_nav() {
+	alcatraz_the_entry_meta( $post_id );
+}
+
+add_action( 'alcatraz_entry_footer_inside', 'alcatraz_output_default_entry_footer' );
```
This still feels a bit messy to me, and wasn't quite what I was expecting. For example, I was trying to customize the footer for one of my child themes, and I still have to write about 30 lines of code to get it to work the way I want.

``` php
add_action( 'alcatraz_before', function() {
    remove_action( 'alcatraz_entry_footer_inside', 'alcatraz_output_default_entry_footer' );
    add_action( 'alcatraz_entry_footer_inside', 'clayton_gerard_entry_footer' );
    /**
     * Customize the entry footer.
     */
    function clayton_gerard_entry_footer( $post_id = 0 ) {

        if ( ! $post_id ) {
            $post_id = get_the_ID();
        }

        $footer_taxonomies = array(
            'categories' => __( 'Posted in: ', 'clayton-gerard' )
        );

        foreach ( $footer_taxonomies as $footer_taxonomy => $label ) {
            alcatraz_the_taxonomy_term_list( $post_id, $footer_taxonomy, $label, ', ' );
        }

        if ( ! is_single() && ! post_password_required() && ( comments_open() || get_comments_number() ) ) {

            echo '<span class="comments-link">';
            comments_popup_link( esc_html__( 'Leave a Comment', 'alcatraz' ), esc_html__( '1 Comment', 'alcatraz' ), esc_html__( '% Comments', 'alcatraz' ) );
            echo '</span>';
        }

        alcatraz_the_edit_post_link( $post_id );
    }
} );
```

And this even isn't working quite right (the categories aren't working). :( I'm concerned that this is too complicated.

BraadMartin committed:
Switch to using display_name over nicename for the author name output

BraadMartin committed:
Remove hr element in entry footers

carrieforde commented:
I appreciate your patience and dedication. Well done, dude. :+1:
